gh-97588: Fix ctypes structs

matthiasgoergens opened the pull request:
The existing code is quite a mess and doesn't correspond to what gcc nor clang nor msvc are doing.

This PR fixes all the issues below.

<!-- gh-issue-number: gh-97588 -->
* Issue: gh-97588
<!-- /gh-issue-number -->

<!-- gh-issue-number: gh-95496 -->
* Issue: gh-95496
<!-- /gh-issue-number -->

<!-- gh-issue-number: gh-84039 -->
* Issue: gh-84039
<!-- /gh-issue-number -->

<!-- gh-issue-number: gh-73939 -->
* Issue: gh-73939
<!-- /gh-issue-number -->

<!-- gh-issue-number: gh-59324 -->
* Issue: gh-59324
<!-- /gh-issue-number -->

<!-- gh-issue-number: gh-86098 -->
* Issue: gh-86098
<!-- /gh-issue-number -->

matthiasgoergens committed:
gh-97588: Fix ctypes structs

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

matthiasgoergens committed:
Handling pack as well

matthiasgoergens committed:
Handle basedict, too

matthiasgoergens committed:
Merge branch 'main' into fix-bitfield-clean

matthiasgoergens committed:
Merge branch 'main' into fix-bitfield-clean

matthiasgoergens committed:
Merge remote-tracking branch 'matthias/fix-bitfield-clean' into fix-bitfield-clean

matthiasgoergens committed:
Merge remote-tracking branch 'origin/main' into fix-bitfield-clean

matthiasgoergens committed:
Split windows and linux

matthiasgoergens committed:
Compiles

matthiasgoergens committed:
Abstract out proto stuff

matthiasgoergens committed:
Clean align

matthiasgoergens committed:
Hypothesis tests pass

matthiasgoergens committed:
Formatting

matthiasgoergens committed:
Clean up

matthiasgoergens committed:
Adapt tests for Windows

matthiasgoergens committed:
Fix order

matthiasgoergens committed:
Fix alignment test

matthiasgoergens committed:
Merge remote-tracking branch 'origin/main' into fix-bitfield-clean

matthiasgoergens committed:
Avoid casting

matthiasgoergens committed:
Fixup

matthiasgoergens committed:
Merge branch 'main' into fix-bitfield-clean

matthiasgoergens committed:
Merge branch 'main' into fix-bitfield-clean

matthiasgoergens committed:
Merge remote-tracking branch 'origin/main' into fix-bitfield-clean

matthiasgoergens committed:
Merge remote-tracking branch 'origin/main' into fix-bitfield-clean

matthiasgoergens committed:
Add ability to force msvc compatibility even when not doing any packing

matthiasgoergens committed:
More tests

matthiasgoergens committed:
Merge branch 'main' into fix-bitfield-clean

matthiasgoergens committed:
Big endian works

matthiasgoergens committed:
Merge remote-tracking branch 'matthias/fix-bitfield-clean' into fix-bitfield-clean

matthiasgoergens committed:
More tests

matthiasgoergens committed:
More asserts

matthiasgoergens committed:
Drop unneeded parameters

matthiasgoergens committed:
Drop unneeded parameters

matthiasgoergens committed:
Drop unneeded parameters

matthiasgoergens committed:
Clean up

matthiasgoergens committed:
Explain

matthiasgoergens committed:
More cleanup

matthiasgoergens committed:
More cleanup

matthiasgoergens committed:
More cleanup

matthiasgoergens committed:
Move common assert

matthiasgoergens committed:
Break line

matthiasgoergens committed:
Clean TODO

matthiasgoergens commented on Modules/_ctypes/stgdict.c:
```diff
@@ -350,11 +350,12 @@ PyCStructUnionType_update_stgdict(PyObject *type, PyObject *fields, int isStruct
     _Py_IDENTIFIER(_swappedbytes_);
     _Py_IDENTIFIER(_use_broken_old_ctypes_structure_semantics_);
     _Py_IDENTIFIER(_pack_);
+    _Py_IDENTIFIER(_ms_struct_);
```
At the moment, this is just so that we can test `msvc` compatibility mode on Linux and vice versa.

We can expose it to users via the docs, too, if we want to.

matthiasgoergens committed:
Merge branch 'main' into fix-bitfield-clean

matthiasgoergens committed:
Fix test for Windows

matthiasgoergens committed:
Merge remote-tracking branch 'matthias/fix-bitfield-clean' into fix-bitfield-clean

matthiasgoergens commented on Modules/_ctypes/cfield.c:
```diff
@@ -147,61 +233,34 @@ PyCField_FromDesc(PyObject *desc, Py_ssize_t index,
+    }
+    assert(bitsize <= dict->size * 8);
+
+    // `pack` only makes sense in msvc compatibility mode.
```
https://docs.python.org/3/library/ctypes.html#structure-union-alignment-and-byte-order says

> By default, Structure and Union fields are aligned in the same way the C compiler does it. It is possible to override this behavior by specifying a _pack_ class attribute in the subclass definition. This must be set to a positive integer and specifies the maximum alignment for the fields. This is what #pragma pack(n) also does in MSVC.

GCC has a suspiciously similarly named `__attribute__((packed))` / `#pragma pack`.  But it does something different.  To get GCC to give you the behaviour as described in the doc, you also need to enable `__attribute__((ms_struct))` / `#pragma ms_struct on`.  And that's what we are doing here.

matthiasgoergens replied:
Note: the behaviour described in the quoted docs that we are implementing here is rather different from what everyone in https://github.com/python/cpython/pull/19850 seems to have understood.

As far as I can tell in that PR everyone assume that `_packed_ = 1` means GCC's `__attribute__((packed))`.

This misunderstanding might be an argument for changing the docs?  However that would be a breaking change.

In general, this PR breaks bug-for-bug compatibility with the past, but does not change documented behaviour.

At the moment, we don't support GCC's version of packed structs.  Neither in the existing code nor in this PR.

It's fairly easy to add this behaviour on top of my code; but I would suggest using a new attribute (like `_gcc_packed_` or so), instead of changing the documented behaviour of the existing `_pack_`.

matthiasgoergens replied:
Correction: support for GCC's packing isn't actually as easy as I imagined.

The problem is that GCC's packing can make eg a `uint8_t` straddle two bytes.  In order to support that, we'd need to revamp much more of ctypes.

matthiasgoergens replied:
Here's an example to show the problem:

```c
#include <stdio.h>
#include <stdint.h>
#include <string.h>

typedef struct
__attribute__ ((packed))
{
  uint8_t a: 4;
  uint8_t b: 8;
  uint8_t c: 4;
} Foo;

typedef union {
  Foo foo;
  uint8_t x[sizeof(Foo)];
} U;

int main(int argc, char** argv) {
    printf("%lu\n", __alignof__(Foo));
    printf("%lu\n", sizeof(Foo));

    U u;
    memset(&u, 0, sizeof(U));
    // Set b to all 1s.
    u.foo.b = -1;
    // Because we are on a little endian machine, it makes more sense to print
    // the bytes in reverse order.
    // This prints "0f f0", showing that b straddles a byte.
    for(int i=sizeof(Foo)-1; i>=0; --i) {
        printf("%02x ", u.x[i]);
    }
    printf("\n");
    return 0;
}
```

encukou replied:
I guess the right thing to do here would be to warn when `_packed_` is used with the GCC semantics.

matthiasgoergens committed:
Remove warning

matthiasgoergens committed:
Port test from PR-19850

matthiasgoergens committed:
Fix assert

matthiasgoergens committed:
Revert "Port test from PR-19850"

This reverts commit ab42187f73dc6d855317cae4587aa6a146b3b1de.

matthiasgoergens committed:
Support gcc's packed structs

matthiasgoergens committed:
Remove gcc_packed experiment

matthiasgoergens committed:
Fix test

matthiasgoergens committed:
Fix ms_struct

matthiasgoergens committed:
Fix ms_struct

matthiasgoergens committed:
Indenting

matthiasgoergens committed:
Merge remote-tracking branch 'origin/main' into fix-bitfield-clean

matthiasgoergens committed:
Merge branch 'main' into fix-bitfield-clean

matthiasgoergens committed:
Fix miscounted parens

matthiasgoergens committed:
Merge branch 'main' into fix-bitfield-clean

matthiasgoergens committed:
Merge remote-tracking branch 'origin/main' into fix-bitfield-clean

matthiasgoergens committed:
make regen-all

matthiasgoergens committed:
clean up

matthiasgoergens committed:
Merge branch 'main' into fix-bitfield-clean

matthiasgoergens committed:
make regen-all

gpshead commented on Lib/test/test_ctypes/test_bitfields.py:
```diff
@@ -51,7 +74,24 @@ def test_shorts(self):
+    def test_shorts_msvc_mode(self):
+        b = BITS_msvc()
+        name = "M"
+        if func_msvc(byref(b), name.encode('ascii')) == 999:
```
add a comment mentioning where the magic 999 comes from (which is already used above, it's just not obvious where that came from for these tests.  (answer: `_ctypes_test.c`)

matthiasgoergens replied:
Good idea.  Done!

gpshead commented on Modules/_ctypes/cfield.c:
```diff
@@ -31,84 +31,170 @@ static void pymem_destructor(PyObject *ptr)
+straddle an alignment boundary for its type, we insert a few bits of padding to
+avoid that.
+
+MSVC mode works similar expect for bitfield packing.  Adjacent bit-fields are
```
s/expect/except/

matthiasgoergens committed:
Merge branch 'main' into fix-bitfield-clean

matthiasgoergens committed:
Merge remote-tracking branch 'origin/main' into fix-bitfield-clean

matthiasgoergens committed:
Explain where 999 comes from

matthiasgoergens committed:
fix typo

matthiasgoergens committed:
Merge remote-tracking branch 'origin/main' into fix-bitfield-clean

matthiasgoergens committed:
Explain magic 999

matthiasgoergens committed:
Incorporate Sam's tests

From https://github.com/python/cpython/pull/103052

matthiasgoergens committed:
Merge remote-tracking branch 'origin/main' into fix-bitfield-clean

matthiasgoergens committed:
Tickle CI/CD

matthiasgoergens committed:
Merge branch 'main' into fix-bitfield-clean

gpshead committed:
Merge branch 'main' into fix-bitfield-clean

matthiasgoergens committed:
Merge commit '698a0da7d440856a90b45964e9082b5a55387b80' into fix-bitfield-clean

matthiasgoergens committed:
Merge commit '0841ca7932987f30a2a23d39f3e6e141622d6fea' into fix-bitfield-clean

matthiasgoergens committed:
Merge commit '9f7176d360b5898003d5ca78bab1822ad67867c4' into fix-bitfield-clean

matthiasgoergens committed:
Fix compile

matthiasgoergens committed:
Merge commit '298bcdc185d1a9709271e61a4cc529d33483add4' into fix-bitfield-clean

matthiasgoergens committed:
Merge commit 'dcaf33a41d5d220523d71c9b35bc08f5b8405dac' into fix-bitfield-clean

matthiasgoergens committed:
Merge commit '7e2fef865899837c47e91ef0180fa59eb03e840b' into fix-bitfield-clean

matthiasgoergens committed:
Merge commit 'ea94b3b149eeadf33c2f7c46f16dcda0adc7cf4e' into fix-bitfield-clean

matthiasgoergens committed:
Merge remote-tracking branch 'origin/main' into fix-bitfield-clean

matthiasgoergens committed:
Fix merge

matthiasgoergens committed:
Clean up

matthiasgoergens committed:
Fix tests

zooba commented on Modules/_ctypes/stgdict.c:
```diff
@@ -271,6 +271,7 @@ PyCStructUnionType_update_stginfo(PyObject *type, PyObject *fields, int isStruct
     if (tmp) {
         pack = PyLong_AsInt(tmp);
         Py_DECREF(tmp);
+        // TODO(Matthias): It looks like pack == 0 triggers a bug.
```
Does this require addressing? Or is it an impossible case

zooba commented on Misc/NEWS.d/next/C API/2022-10-01-09-56-27.gh-issue-97588.Gvg54o.rst:
```diff
@@ -0,0 +1 @@
+Fix ctypes construction of structs from description.
```
Can you expand a bit here? This change could presumably impact anyone using packed bit fields through ctypes, so it would be good to mention at least some of those keywords here. A couple of sentences are okay (and feel free to add "`Contributed by <your name>"` at the end if you like - you've earned it 😉 )

gvanrossum replied:
In fact maybe a what's new entry would be appreciated too?

zooba replied:
Maybe... "we're getting it right now" isn't the sort of thing we'd usually advertise as "new", but on the other hand, there's a good chance people have discovered it was buggy before and are now avoiding it. A note there may get noticed.

We still have someone edit what's new before it goes out, right? It won't hurt to add it, and the editor can choose whether to keep it or rephrase/reframe it.

For @matthiasgoergens, we're talking about `Doc/whatsnew/3.13.rst` as I mentioned for porting notes, but in the "Improved Modules" section (`ctypes` probably needs to be added as a heading).

zooba reviewed (dismissed):
I'll be honest that I don't fully understand how the changes work, but we're okay to accept the changes based on good test coverage. It would be great if you could keep an eye out for any issues that arise, as we may need your help again to deal with them.

I'm also concerned about the _subtlety_ of changes to existing behaviour, and so I don't want to backport this to earlier releases. That way we have a clear boundary where any workarounds become unnecessary.

It would also be good to add a short explanation to the Porting Notes (`Doc\whatsnew\3.13.rst` section titled "Porting to Python 3.13"). That will help get the attention of anyone who may need to adjust their existing definitions to handle the correct behaviour.

gpshead commented on Lib/test/test_ctypes/test_bitfields.py:
```diff
@@ -42,18 +67,43 @@ def test_ints(self):
+                self.assertEqual(
+                    getattr(b, name),
+                    func(byref(b), (name.encode('ascii'))),
+                    (name, i))
```
consider updating tests like this that loop over byte values into sub-tests via the https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests API call so that you don't need to pass `msg=(name, i)` here to get a meaningful error message.

gpshead reviewed (commented):
Overall I think this is a great improvement.  I didn't dive into the details of the implementation - mostly looking at it from the level of tests and concept.  I believe @encukou will be taking a closer look soon.

encukou commented on Modules/_ctypes/_ctypes_test.c:
```diff
@@ -734,6 +759,49 @@ EXPORT(int) unpack_bitfields(struct BITS *bits, char name)
 
+struct
+#ifndef MS_WIN32
+__attribute__ ((ms_struct))
```
`__attribute__` is a GCC extension. (Also supported by clang, but I'm not immediately sure if clang has `ms_struct`.)
It should be guarded by something like `defined(__GNUC__)`, and the tests skipped on unknown platforms.

encukou commented on Lib/test/test_ctypes/test_bitfields.py:
```diff
@@ -236,6 +286,157 @@ class X(Structure):
+                ("A", c_uint),
+                ('B', c_uint, 20),
+                ('C', c_ulonglong, 24)]
+        self.assertEqual(16, sizeof(X))
```
This assumes specific `sizeof(c_uint)` & `sizeof(c_ulonglong)`.

encukou commented on Modules/_ctypes/cfield.c:
```diff
@@ -31,40 +31,159 @@ static void pymem_destructor(PyObject *ptr)
+See https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html#index-mms-bitfields for details.
+
+We do not support zero length bitfields.  In fact we use bitsize != 0 to
+indicate a bitfield.
```
You do add a separate `is_bitfield` argument. Please use it consistently, with
```suggestion
We do not support zero length bitfields.  In fact we use bitsize != 0 elsewhere to
indicate a bitfield.
```
or remove it.

encukou commented on Modules/_ctypes/cfield.c:
```diff
@@ -31,40 +31,159 @@ static void pymem_destructor(PyObject *ptr)
+PyCField_FromDesc_gcc(Py_ssize_t bitsize, Py_ssize_t *pbitofs,
                 Py_ssize_t *psize, Py_ssize_t *poffset, Py_ssize_t *palign,
-                int pack, int big_endian)
+                CFieldObject* self, StgInfo* dict,
```
When I changed StgDict to StgInfo, it took several commits (and false attempts); to keep it all straight I renamed `dict` variables to `info`.
I recommend doing it too, so the names aren't misleading.
```suggestion
                CFieldObject* self, StgInfo* info,
```
Sorry for the extra work :(

(For the record: When there's a need to access `__dict__` (attributes of the Python object), I called the new variable `attrdict`. Nowadays, after the transition, `dict` is OK for that.)

encukou commented on Modules/_ctypes/cfield.c:
```diff
@@ -31,40 +31,159 @@ static void pymem_destructor(PyObject *ptr)
+        self->offset = round_down(*pbitofs, 8*dict->size) / 8;
+        Py_ssize_t effective_bitsof = *pbitofs - 8 * self->offset;
+        self->size = BUILD_SIZE(bitsize, effective_bitsof);
+        assert(dict->size == dict->align);
```
Hm, this restriction deserves a comment at least. I guess all supported platforms have the property, but I'm currently not aware of it being required or tested for CPython.

Perhaps rather than a debug assertion, it should be a `if` and raise an exception.

encukou commented on Modules/_ctypes/cfield.c:
```diff
@@ -31,40 +31,159 @@ static void pymem_destructor(PyObject *ptr)
+    if (pack)
+        *palign = min(pack, dict->align);
+    else
+        *palign = dict->align;
```
* PEP-7 nitpick: use braces for all blocks.

* AFAIK, the `min` define is a ctypes-specific accident of history; please use CPython's `Py_MIN`.

```suggestion
    if (pack) {
        *palign = Py_MIN(pack, dict->align);
    } else {
        *palign = dict->align;
    }
```

(I see this is moved, rather than new code. Still, probably better to touch it up.)

encukou reviewed (commented):
Thank you for taking this ~~minefield~~ bitfield on! There's a bunch of careful work to do, alas.
I do agree that it would be great to get this into 3.13, even if further work remains for 3.14.

I would love to run those Hypothesis tests on as many platforms as possible. Could you write up some pointers on where to get them and how to run them?

I attached some review comments. I'm not done reviewing yet -- I now just have a good idea of how the pieces fit together :)
But I reckon it's good to share these early.

---

Some thoughts on possible future direction:

Zooming out: in terms of C standard, bitfield packing is implementation-defined: this means the compiler decides about the details; so we have GCC semantics and MSVC semantics -- covering the two "platforms" people care about. However, GCC [delegates *some* of the details to “the ABI”](https://gcc.gnu.org/onlinedocs/gcc-13.2.0/gcc/Structures-unions-enumerations-and-bit-fields-implementation.html), so it'd be more correct to call the semantics "gcc-x86-64" rather than just "gcc".
(In practice, portable programs avoid bitfields, so there's not too much incentive to standardize, and not *too* much need for ctypes or libffi to support them. But, if the plan is to rewrite the logic in Python, hopefully it'll become easier for users to plug other behaviour in.)
(Relatedly, making bitfields out of anything other than [signed|unsigned] int and _Bool is an implementation-specific extension. In that `SIGNED_SHORT_BITFIELDS` special case, the `__xlc__` compiler isn't doing anything wrong.)

Eventually, I think we need to re-think the ctypes design decision to try *only* match whatever compiler (and compiler settings) CPython was compiled with.
That's hard to get right, and I also doubt that it's all that useful for *users* when they get into a situation where bit field packing matters. And behaviour that's essentially based on a guess is hard to maintain, especially in a backwards-compatible way.

I'd love to see a user-settable `__packing_strategy__` attribute, with one place that sets it to "gcc-x86-64" or "ms" in case it's not specified. Most of the code can then branch on that, rather than continuing “Windows vs. anything else”.
Then, the bulk of the testing could be done on any platform; only comparison to compiler output would be OS-specific. (Preferably the tests would run any available compiler(s) directly, so they won't fail when you happen to compile Python with `CFLAGS=-mms-bitfields`.)

But, baby steps. The current scope of this PR is good for an improvement to 3.13. It does need to be validated on (and corrected for) more platforms, though.

encukou committed:
Merge in the main branch

encukou committed:
Add Hypothesis test

encukou committed:
Merge in the main branch

encukou committed:
dict -> info

encukou committed:
Conditionalize ms_struct

encukou committed:
Use subTest

encukou committed:
Use fixed-width types in tests

encukou committed:
fixup! dict

encukou committed:
Turn `info->size != info->align` assertion into a TypeError

encukou committed:
PEP 7

encukou committed:
Change _ms_struct_ to _layout_

encukou committed:
Merge in the main branch

encukou committed:
PEP 7

encukou committed:
TMP

encukou committed:
Add generated tests

encukou committed:
Merge in the main branch

encukou committed:
Remove ideas

encukou committed:
Regen

encukou committed:
Do not set _layout_

encukou committed:
We don't always get union alignment right

encukou committed:
Remove unused stuff

encukou committed:
Packing implies ms layout

encukou committed:
Docs & whitespace

encukou committed:
Use limited API again

encukou committed:
Appease the linter

encukou committed:
Fix ReST typo

encukou committed:
Use nicer error output for memory dumps

encukou committed:
Regen global strings

encukou committed:
Allow alignment < size, as in int64_t on x86 (32-bit), GCC layout

encukou committed:
Add examples from MS docs

encukou committed:
Use correct spelling for test skips

encukou committed:
Only use 'attribute((ms_struct))' on x86_64 & ppc; on GCC and clang

encukou committed:
Remove the `info->size == info->align` assertion for _layout_='ms'

The MS compiler aligns data based on its size by default, but the 'ms'
layout might be used with other ABIs.

encukou committed:
Fix silly mistake in the macro

encukou committed:
Include field name in error message

encukou committed:
Regen

encukou committed:
Conditionally skip tests that assume sizeof(int64) == alignment(int64)

encukou committed:
Use ms_struct only on Windows and x86/ppc GCC/clang

encukou committed:
Use proper PyArg_ParseTuple code for Py_ssize_t

encukou committed:
Merge in the main branch
